photon-client/src/main/scala/com/linkedin/photon/ml/data/DataValidators.scala

```diff
@@ -259,26 +259,26 @@ object DataValidators extends Logging {
       perSampleValidators: List[(((Row, String) => Boolean), InputColumnsNames.Value, String)],
       inputColumnsNames: InputColumnsNames,
       featureSectionKeys: Set[FeatureShardId]): Seq[String] = {
-
-    perSampleValidators
-      .flatMap { case (validator, columnName, msg) =>
-        val rows = dataSet.collect()
-        rows.flatMap { r =>
-          if (columnName.equals(InputColumnsNames.FEATURES_DEFAULT)) {
-            featureSectionKeys.map(features => (validator(r, features), msg))
+    
+    val columns = dataSet.columns
+    dataSet.rdd.flatMap { r =>
+      perSampleValidators.flatMap { case (validator, columnName, msg) =>
+        if (columnName.equals(InputColumnsNames.FEATURES_DEFAULT)) {
+          featureSectionKeys.map(features => (validator(r, features), msg))
```

Comment:
It looks like if the validator is specified for the default feature column name, we run validators for all feature columns in one pass. What's the reason for this?

Reply:
The if/else block here separates the case where we validate the "features" column from other columns. Other columns have only one field to validate (e.g. "response"), while for the "features" column we'd like to validate for all the feature shard ids.